bindings/pyroot_experimental/PyROOT/src/TDirectoryPyz.cxx

```diff
@@ -0,0 +1,113 @@
+#include "PyzCppHelpers.hxx"
+#include "TClass.h"
+#include "TDirectory.h"
+#include "TKey.h"
+#include "Python.h"
+
+using namespace CPyCppyy;
+
+////////////////////////////////////////////////////////////////////////////////
+/// \brief Implements the WriteObject method of TDirectory
+/// This method allows to write objects into TDirectory instances with this
+/// syntax:
+/// ~~~{.python}
+/// myDir.WriteObject(myObj, "myKeyName")
+/// ~~~
+PyObject *TDirectoryWriteObject(CPPInstance *self, PyObject *args)
+{
+   CPPInstance *wrt = nullptr;
+   PyObject *name = nullptr;
+   PyObject *option = nullptr;
+   Int_t bufsize = 0;
+   if (!PyArg_ParseTuple(args, const_cast<char *>("O!O!|O!i:TDirectory::WriteObject"), &CPPInstance_Type, &wrt,
+                         &CPyCppyy_PyUnicode_Type, &name, &CPyCppyy_PyUnicode_Type, &option, &bufsize))
+      return nullptr;
+   auto dir = (TDirectory *)OP2TCLASS(self)->DynamicCast(TDirectory::Class(), self->GetObject());
+   if (!dir) {
+      PyErr_SetString(PyExc_TypeError,
+                      "TDirectory::WriteObject must be called with a TDirectory instance as first argument");
+      return nullptr;
+   }
+   Int_t result = 0;
+   if (option != nullptr) {
+      result = dir->WriteObjectAny(wrt->GetObject(), OP2TCLASS(wrt), CPyCppyy_PyUnicode_AsString(name),
+                                   CPyCppyy_PyUnicode_AsString(option), bufsize);
+   } else {
+      result = dir->WriteObjectAny(wrt->GetObject(), OP2TCLASS(wrt), CPyCppyy_PyUnicode_AsString(name));
+   }
+   return PyInt_FromLong((Long_t)result);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// \brief Implement the Get method for TDirectory
+/// This allows to seamlessly read from a TDirectory, and therefore a TDirectoryFile
+/// and file objects inputting their key name, being them TObjects or not.
+PyObject *TDirectoryGet(CPPInstance *self, PyObject *pynamecycle)
+{
+   // Pythonization of TDirectory::Get that handles non-TObject deriveds
+   if (!CPPInstance_Check(self)) {
+      PyErr_SetString(PyExc_TypeError, "TDirectory::Get must be called with a TDirectory instance as first argument");
+      return nullptr;
+   }
+   auto dirf = (TDirectory *)OP2TCLASS(self)->DynamicCast(TDirectory::Class(), self->GetObject());
+   if (!dirf) {
+      PyErr_SetString(PyExc_ReferenceError, "attempt to access a null-pointer");
+      return nullptr;
+   }
+   const char *namecycle = CPyCppyy_PyUnicode_AsString(pynamecycle);
+   if (!namecycle)
+      return nullptr; // TypeError already set
+   auto key = dirf->GetKey(namecycle);
+   if (key) {
+      void *addr = dirf->GetObjectChecked(namecycle, key->GetClassName());
+      return BindCppObjectNoCast(addr, (Cppyy::TCppType_t)Cppyy::GetScope(key->GetClassName()), kFALSE);
+   }
+   // no key? for better or worse, call normal Get()
+   void *addr = dirf->Get(namecycle);
+   return BindCppObject(addr, (Cppyy::TCppType_t)Cppyy::GetScope("TObject"), kFALSE);
```

Comment:
maybe mention that BindCppObject is able to cast to the right type since we know this is a TObject since you can only store tobjects in TDirectory.